node/full_node_integration_test.go

```diff
@@ -406,6 +408,89 @@ func TestSubmitBlocksToDA(t *testing.T) {
 	}
 }
 
+func TestMaxPending(t *testing.T) {
+	cases := []struct {
+		name       string
+		maxPending uint64
+	}{
+		{
+			name:       "no limit",
+			maxPending: 0,
+		},
+		{
+			name:       "10 pending blocks limit",
+			maxPending: 10,
+		},
+		{
+			name:       "50 pending blocks limit",
+			maxPending: 50,
+		},
+	}
+
+	for _, tc := range cases {
+		t.Run(tc.name, func(t *testing.T) {
+			doTestMaxPending(tc.maxPending, t)
+		})
+	}
+}
+
+func doTestMaxPending(maxPending uint64, t *testing.T) {
+	require := require.New(t)
+
+	clientNodes := 1
+	ctx, cancel := context.WithTimeout(context.Background(), 3*time.Second)
+	defer cancel()
+	nodes, _ := createNodes(
+		ctx,
+		context.Background(),
+		clientNodes,
+		config.BlockManagerConfig{
+			DABlockTime:      20 * time.Millisecond,
+			BlockTime:        10 * time.Millisecond,
+			MaxPendingBlocks: maxPending,
+		},
+		t,
+	)
+	seq := nodes[0]
+	mockDA := &mocks.DA{}
+
+	// make sure mock DA is not accepting any submissions
+	mockDA.On("MaxBlobSize", mock.Anything).Return(uint64(123456789), nil)
+	mockDA.On("Submit", mock.Anything, mock.Anything, mock.Anything, mock.Anything).Return(nil, errors.New("DA not available"))
+
+	dalc := da.NewDAClient(mockDA, 1234, 5678, goDA.Namespace(MockDANamespace), log.NewNopLogger())
+	require.NotNil(dalc)
+	seq.dalc = dalc
+	seq.blockManager.SetDALC(dalc)
+
+	startNodeWithCleanup(t, seq)
+
+	if maxPending == 0 { // if there is no limit, sequencer should produce blocks even DA is unavailable
+		require.NoError(waitForAtLeastNBlocks(seq, 3, Store))
+		return
+	} else { // if there is a limit, sequencer should produce exactly maxPending blocks and pause
+		require.NoError(waitForAtLeastNBlocks(seq, int(maxPending), Store))
+		// wait few block times and ensure that new blocks are not produced
+		time.Sleep(3 * seq.nodeConfig.BlockTime)
+		require.EqualValues(maxPending, seq.Store.Height())
```

Comment:
Are there any alternatives to using a sleep here?

Wonder what @MSevey thinks about this `Sleep` statement as well.

Reply:
As we're looking for "not producing" and idling for some period of time, it's hard to think about any other way than just sleep. Block times are extremely short, so this shouldn't be long wait.

Reply:
I think this is ok, because we are confirming non action. 
Most of the scope of removing bare sleeps with more dynamic solutions like retries, was confirming an action that we weren't sure when it would happen. In which case the sleeps were guesses.